File: apps/api/src/routers/cms/news.py
from fastapi import APIRouter, Depends, File, Query, Request, UploadFile
from sqlmodel.ext.asyncio.session import AsyncSession

from src.core.events.database import get_db_session
from src.db.cms_news import (
    CMSNewsCreate,
    CMSNewsListResponse,
    CMSNewsRead,
    CMSNewsUpdate,
)
from src.db.users import PublicUser
from src.security.auth import get_current_user
from src.services.cms import news as news_service

router = APIRouter()


@router.get(
    "/org/{org_slug}/news",
    response_model=CMSNewsListResponse,
    summary="List published news for an organization",
    description="Public list of published news items (latest first). Body is omitted.",
)
async def api_list_public_news(
    request: Request,
    org_slug: str,
    page: int = Query(1, ge=1),
    limit: int = Query(12, ge=1, le=100),
    current_user: PublicUser = Depends(get_current_user),
    db_session: AsyncSession = Depends(get_db_session),
) -> CMSNewsListResponse:
    return await news_service.list_public_news(
        request, org_slug, page, limit, current_user, db_session
    )


@router.get(
    "/org/{org_slug}/news/{slug}",
    response_model=CMSNewsRead,
    summary="Get a published news article by slug",
)
async def api_get_public_news(
    request: Request,
    org_slug: str,
    slug: str,
    current_user: PublicUser = Depends(get_current_user),
    db_session: AsyncSession = Depends(get_db_session),
) -> CMSNewsRead:
    return await news_service.get_public_news_by_slug(
        request, org_slug, slug, current_user, db_session
    )


@router.get(
    "/admin/{org_id}/news",
    response_model=CMSNewsListResponse,
    summary="List all news for an organization (admin)",
)
async def api_list_admin_news(
    request: Request,
    org_id: int,
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    current_user: PublicUser = Depends(get_current_user),
    db_session: AsyncSession = Depends(get_db_session),
) -> CMSNewsListResponse:
    return await news_service.list_admin_news(
        request, org_id, page, limit, current_user, db_session
    )


@router.get(
    "/admin/{org_id}/news/{news_uuid}",
    response_model=CMSNewsRead,
    summary="Get a news article by UUID (admin)",
)
async def api_get_admin_news(
    request: Request,
    org_id: int,
    news_uuid: str,
    current_user: PublicUser = Depends(get_current_user),
    db_session: AsyncSession = Depends(get_db_session),
) -> CMSNewsRead:
    return await news_service.get_admin_news(
        request, org_id, news_uuid, current_user, db_session
    )


@router.post(
    "/admin/{org_id}/news",
    response_model=CMSNewsRead,
    summary="Create a news article",
)
async def api_create_news(
    request: Request,
    org_id: int,
    payload: CMSNewsCreate,
    current_user: PublicUser = Depends(get_current_user),
    db_session: AsyncSession = Depends(get_db_session),
) -> CMSNewsRead:
    return await news_service.create_news(
        request, org_id, payload, current_user, db_session
    )


@router.put(
    "/admin/{org_id}/news/{news_uuid}",
    response_model=CMSNewsRead,
    summary="Update a news article",
)
async def api_update_news(
    request: Request,
    org_id: int,
    news_uuid: str,
    payload: CMSNewsUpdate,
    current_user: PublicUser = Depends(get_current_user),
    db_session: AsyncSession = Depends(get_db_session),
) -> CMSNewsRead:
    return await news_service.update_news(
        request, org_id, news_uuid, payload, current_user, db_session
    )


@router.post(
    "/admin/{org_id}/news/{news_uuid}/cover",
    response_model=CMSNewsRead,
    summary="Upload a news cover image",
)
async def api_upload_news_cover(
    request: Request,
    org_id: int,
    news_uuid: str,
    cover_file: UploadFile = File(...),
    current_user: PublicUser = Depends(get_current_user),
    db_session: AsyncSession = Depends(get_db_session),
) -> CMSNewsRead:
    return await news_service.upload_news_cover(
        request, org_id, news_uuid, cover_file, current_user, db_session
    )


@router.post(
    "/admin/{org_id}/news/{news_uuid}/images",
    response_model=CMSNewsRead,
    summary="Upload an image attachment to a news article",
)
async def api_upload_news_image(
    request: Request,
    org_id: int,
    news_uuid: str,
    image_file: UploadFile = File(...),
    current_user: PublicUser = Depends(get_current_user),
    db_session: AsyncSession = Depends(get_db_session),
) -> CMSNewsRead:
    return await news_service.upload_news_image(
        request, org_id, news_uuid, image_file, current_user, db_session
    )


@router.post(
    "/admin/{org_id}/news/{news_uuid}/videos",
    response_model=CMSNewsRead,
    summary="Upload a video attachment to a news article",
)
async def api_upload_news_video(
    request: Request,
    org_id: int,
    news_uuid: str,
    video_file: UploadFile = File(...),
    current_user: PublicUser = Depends(get_current_user),
    db_session: AsyncSession = Depends(get_db_session),
) -> CMSNewsRead:
    return await news_service.upload_news_video(
        request, org_id, news_uuid, video_file, current_user, db_session
    )


@router.delete(
    "/admin/{org_id}/news/{news_uuid}/images/{media_id}",
    response_model=CMSNewsRead,
    summary="Remove an image attachment from a news article",
)
async def api_delete_news_image(
    request: Request,
    org_id: int,
    news_uuid: str,
    media_id: int,
    current_user: PublicUser = Depends(get_current_user),
    db_session: AsyncSession = Depends(get_db_session),
) -> CMSNewsRead:
    return await news_service.delete_news_image(
        request, org_id, news_uuid, media_id, current_user, db_session
    )


@router.delete(
    "/admin/{org_id}/news/{news_uuid}/videos/{media_id}",
    response_model=CMSNewsRead,
    summary="Remove a video attachment from a news article",
)
async def api_delete_news_video(
    request: Request,
    org_id: int,
    news_uuid: str,
    media_id: int,
    current_user: PublicUser = Depends(get_current_user),
    db_session: AsyncSession = Depends(get_db_session),
) -> CMSNewsRead:
    return await news_service.delete_news_video(
        request, org_id, news_uuid, media_id, current_user, db_session
    )


@router.delete(
    "/admin/{org_id}/news/{news_uuid}",
    summary="Delete a news article",
)
async def api_delete_news(
    request: Request,
    org_id: int,
    news_uuid: str,
    current_user: PublicUser = Depends(get_current_user),
    db_session: AsyncSession = Depends(get_db_session),
):
    return await news_service.delete_news(
        request, org_id, news_uuid, current_user, db_session
    )

File: apps/web/services/cms/news.ts
import { getAPIUrl } from '@services/config/config'
import {
  RequestBodyWithAuthHeader,
  RequestBodyFormWithAuthHeader,
  errorHandling,
} from '@services/utils/ts/requests'

export type CMSNewsImage = {
  id: number
  imageName: string
  imageURL: string
  dateCreatedString: string
}

export type CMSNewsVideo = {
  id: number
  videoName: string
  videoURL: string
  dateCreatedString: string
}

export type CMSNewsListItem = {
  id: number
  org_id: number
  org_uuid?: string
  news_uuid: string
  title: string
  slug: string
  excerpt: string
  cover_image: string
  images: CMSNewsImage[]
  videos: CMSNewsVideo[]
  published: boolean
  published_at: string | null
  creation_date: string
  update_date: string
}

export type CMSNews = CMSNewsListItem & {
  body: string
  created_by?: number | null
}

export type CMSNewsListResponse = {
  items: CMSNewsListItem[]
  total: number
  page: number
  limit: number
}

export type CMSNewsPayload = {
  title: string
  slug?: string | null
  excerpt?: string
  body?: string
  cover_image?: string
  published?: boolean
  published_at?: string | null
}

export async function listAdminNews(
  org_id: number,
  access_token: string,
  page = 1,
  limit = 20
): Promise<CMSNewsListResponse> {
  const result = await fetch(
    `${getAPIUrl()}cms/admin/${org_id}/news?page=${page}&limit=${limit}`,
    RequestBodyWithAuthHeader('GET', null, null, access_token)
  )
  return errorHandling(result)
}

export async function getAdminNews(
  org_id: number,
  news_uuid: string,
  access_token: string
): Promise<CMSNews> {
  const result = await fetch(
    `${getAPIUrl()}cms/admin/${org_id}/news/${news_uuid}`,
    RequestBodyWithAuthHeader('GET', null, null, access_token)
  )
  return errorHandling(result)
}

export async function createNews(
  org_id: number,
  data: CMSNewsPayload,
  access_token: string
): Promise<CMSNews> {
  const result = await fetch(
    `${getAPIUrl()}cms/admin/${org_id}/news`,
    RequestBodyWithAuthHeader('POST', data, null, access_token)
  )
  return errorHandling(result)
}

export async function updateNews(
  org_id: number,
  news_uuid: string,
  data: Partial<CMSNewsPayload>,
  access_token: string
): Promise<CMSNews> {
  const result = await fetch(
    `${getAPIUrl()}cms/admin/${org_id}/news/${news_uuid}`,
    RequestBodyWithAuthHeader('PUT', data, null, access_token)
  )
  return errorHandling(result)
}

export async function deleteNews(
  org_id: number,
  news_uuid: string,
  access_token: string
): Promise<{ detail: string }> {
  const result = await fetch(
    `${getAPIUrl()}cms/admin/${org_id}/news/${news_uuid}`,
    RequestBodyWithAuthHeader('DELETE', null, null, access_token)
  )
  return errorHandling(result)
}

export async function uploadNewsCover(
  org_id: number,
  news_uuid: string,
  file: File,
  access_token: string
): Promise<CMSNews> {
  const formData = new FormData()
  formData.append('cover_file', file)
  const result = await fetch(
    `${getAPIUrl()}cms/admin/${org_id}/news/${news_uuid}/cover`,
    RequestBodyFormWithAuthHeader('POST', formData, null, access_token)
  )
  return errorHandling(result)
}

export async function uploadNewsImage(
  org_id: number,
  news_uuid: string,
  file: File,
  access_token: string
): Promise<CMSNews> {
  const formData = new FormData()
  formData.append('image_file', file)
  const result = await fetch(
    `${getAPIUrl()}cms/admin/${org_id}/news/${news_uuid}/images`,
    RequestBodyFormWithAuthHeader('POST', formData, null, access_token)
  )
  return errorHandling(result)
}

export async function uploadNewsVideo(
  org_id: number,
  news_uuid: string,
  file: File,
  access_token: string
): Promise<CMSNews> {
  const formData = new FormData()
  formData.append('video_file', file)
  const result = await fetch(
    `${getAPIUrl()}cms/admin/${org_id}/news/${news_uuid}/videos`,
    RequestBodyFormWithAuthHeader('POST', formData, null, access_token)
  )
  return errorHandling(result)
}

export async function deleteNewsImage(
  org_id: number,
  news_uuid: string,
  media_id: number,
  access_token: string
): Promise<CMSNews> {
  const result = await fetch(
    `${getAPIUrl()}cms/admin/${org_id}/news/${news_uuid}/images/${media_id}`,
    RequestBodyWithAuthHeader('DELETE', null, null, access_token)
  )
  return errorHandling(result)
}

export async function deleteNewsVideo(
  org_id: number,
  news_uuid: string,
  media_id: number,
  access_token: string
): Promise<CMSNews> {
  const result = await fetch(
    `${getAPIUrl()}cms/admin/${org_id}/news/${news_uuid}/videos/${media_id}`,
    RequestBodyWithAuthHeader('DELETE', null, null, access_token)
  )
  return errorHandling(result)
}
